File: php/test/CollaboratorEntityTest.php
<?php
declare(strict_types=1);

// Collaborator entity test

require_once __DIR__ . '/../gamedevelopment_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class CollaboratorEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = GameDevelopmentSDK::test(null, null);
        $ent = $testsdk->Collaborator(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = collaborator_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "collaborator." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set GAMEDEVELOPMENT_TEST_COLLABORATOR_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $collaborator_ref01_ent = $client->Collaborator(null);
        $collaborator_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.collaborator"), "collaborator_ref01"));
        $collaborator_ref01_data["project_id"] = $setup["idmap"]["project01"];

        [$collaborator_ref01_data_result, $err] = $collaborator_ref01_ent->create($collaborator_ref01_data, null);
        $this->assertNull($err);
        $collaborator_ref01_data = Helpers::to_map($collaborator_ref01_data_result);
        $this->assertNotNull($collaborator_ref01_data);

    }
}

function collaborator_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/collaborator/CollaboratorTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = GameDevelopmentSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["collaborator01", "collaborator02", "collaborator03", "project01", "project02", "project03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("GAMEDEVELOPMENT_TEST_COLLABORATOR_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "GAMEDEVELOPMENT_TEST_COLLABORATOR_ENTID" => $idmap,
        "GAMEDEVELOPMENT_TEST_LIVE" => "FALSE",
        "GAMEDEVELOPMENT_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["GAMEDEVELOPMENT_TEST_COLLABORATOR_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["GAMEDEVELOPMENT_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new GameDevelopmentSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["GAMEDEVELOPMENT_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["GAMEDEVELOPMENT_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: go/test/collaborator_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/game-development-sdk/go"
	"github.com/voxgig-sdk/game-development-sdk/go/core"

	vs "github.com/voxgig-sdk/game-development-sdk/go/utility/struct"
)

func TestCollaboratorEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Collaborator(nil)
		if ent == nil {
			t.Fatal("expected non-nil CollaboratorEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := collaboratorBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "collaborator." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GAME_DEVELOPMENT_TEST_COLLABORATOR_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		collaboratorRef01Ent := client.Collaborator(nil)
		collaboratorRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "collaborator"}), "collaborator_ref01"))
		collaboratorRef01Data["project_id"] = setup.idmap["project01"]

		collaboratorRef01DataResult, err := collaboratorRef01Ent.Create(collaboratorRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		collaboratorRef01Data = core.ToMapAny(entityData(collaboratorRef01DataResult))
		if collaboratorRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

	})
}

func collaboratorBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "collaborator", "CollaboratorTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read collaborator test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse collaborator test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"collaborator01", "collaborator02", "collaborator03", "project01", "project02", "project03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GAME_DEVELOPMENT_TEST_COLLABORATOR_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GAME_DEVELOPMENT_TEST_COLLABORATOR_ENTID": idmap,
		"GAME_DEVELOPMENT_TEST_LIVE":      "FALSE",
		"GAME_DEVELOPMENT_TEST_EXPLAIN":   "FALSE",
		"GAME_DEVELOPMENT_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["GAME_DEVELOPMENT_TEST_COLLABORATOR_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GAME_DEVELOPMENT_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["GAME_DEVELOPMENT_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewGameDevelopmentSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GAME_DEVELOPMENT_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GAME_DEVELOPMENT_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
